diff --git a/.htaccess b/.htaccess
index 6f9123d..84ec3c6 100644
--- a/.htaccess
+++ b/.htaccess
@@ -102,8 +102,8 @@ AddEncoding gzip svgz
   # To redirect all users to access the site WITHOUT the 'www.' prefix,
   # (http://www.example.com/foo will be redirected to http://example.com/foo)
   # uncomment the following:
-  # RewriteCond %{HTTP_HOST} ^www\.(.+)$ [NC]
-  # RewriteRule ^ http%{ENV:protossl}://%1%{REQUEST_URI} [L,R=301]
+  RewriteCond %{HTTP_HOST} ^www\.(.+)\.lndo\.site$ [NC]
+  RewriteRule ^ http%{ENV:protossl}://%1.lndo.site%{REQUEST_URI} [L,R=301]

   # Modify the RewriteBase if you are using Drupal in a subdirectory or in a
   # VirtualDocumentRoot and the rewrite rules are not working properly.
